lib/presets.js

```diff
@@ -305,22 +305,67 @@ module.exports = {
      * });
      */
     cargo: (options, cwd, version, callback) => {
-      const cargoFiles = path.join(cwd, 'Cargo.*');
+      const cargoToml = path.join(cwd, 'Cargo.toml');
+      const cargoLock = path.join(cwd, 'Cargo.lock');
+
       const cleanedVersion = semver.clean(version);
 
       if (!cleanedVersion) {
         return callback(new Error(`Invalid version: ${version}`));
       }
 
-      replace({
-        files: cargoFiles,
-        from: /(version\s*=\s*)"(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)"/g,
-        to: '$1"' + cleanedVersion + '"'
-      }, (error) => {
-        if (error) {
-          return callback(error);
+      async.waterfall([
+        (done) => {
+          return fs.readFile(cargoToml, 'utf8', done);
+        },
+
+        (contents, done) => {
+          // Capture first `name = "..."` occurrence immediately after `[package]`
+          const matches = contents.match(/\[package\][^]+?name\s*=\s*("|')(.+?)\1/m);
+          if (_.isNull(matches)) {
+            done(new Error(`Package name not found in ${cargoToml}`));
+          } else {
+            done(null, matches[2]);
+          }
+        },
+
+        (packageName, done) => {
+          if (fs.existsSync(cargoLock)) {
+            // Update first `version = "..."` occurrence immediately after `name = "${packageName}"`
+            const cargoLockRe = new RegExp(
+              `(name\\s*=\\s*(?:"|')${packageName}(?:"|')[^]+?version\\s*=\\s*)("|').*?\\2`, 'm'
+            );
+            replace({
+              files: cargoLock,
+              from: cargoLockRe,
+              to: '$1$2' + cleanedVersion + '$2'
+            }, done);
+          } else {
+            done(null, null);
+          }
+        },
+
+        (changedFiles, done) => {
+          if (_.isArray(changedFiles) && _.isEmpty(changedFiles)) {
+            return done(new Error(`Updating ${cargoLock} failed`));
+          }
+
+          // Update first `version = "..."` occurrence immediately after `[package]`
+          replace({
+            files: cargoToml,
+            from: /(\[package\][^]+?version\s*=\s*)("|').*?\2/m,
+            to: '$1$2' + cleanedVersion + '$2'
+          }, done);
+        },
+
+        (changedFiles, done) => {
+          if (_.isArray(changedFiles) && _.isEmpty(changedFiles)) {
```

Comment:
It obviously doesn't hurt anything, but I guess _strictly_ speaking the `_.isArray` isn't needed here?

Reply:
In the case above this one `_.isArray` is needed, because we want to ensure that the passed value is an empty array (`_.isEmpty(null)` is true in the case above). I placed it here as well for consistency. It is not needed here though in terms of correctness as you mention.

The problem is that in the first case `changedFiles` can be both a `null` and an array. In the second: an array only.

Maybe it is better to change `_.isArray(changedFiles)` to `!_.isNull(changedFiles)`, which describes better what is happening.

What's your opinion on all of this?

Reply:
I'm happy for it to be left as-is ("doesn't hurt anything"), I was merely asking to check that I understood exactly what the code was doing ("In the second: an array only"), and you've confirmed that I have :-)

Reply:
Leaving it as is.